### DISCUSSION TITLE
Watchdog timer triggered on MBsrv01RTU

### DISCUSSION THREAD

**Discussion by mstegen** [eModbus/eModbus#120] [Q&A]
Hi again, 
I'm porting some more code to use the eModbus library.

This project sometimes (happened twice so far) resets with a 

```
E (1267264) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (1267264) task_wdt:  - IDLE0 (CPU 0)
E (1267264) task_wdt: Tasks currently running:
E (1267264) task_wdt: CPU 0: MBsrv01RTU
E (1267264) task_wdt: CPU 1: loopTask
E (1267264) task_wdt: Aborting.
abort() was called at PC 0x4014b100 on core 0
```

It uses all tree Uarts of the ESP32, Uart2 receives a burst (~1000 bytes) of data @ 115200bps each second.
eModbus RTU server (9600bps) is on Uart1.
Uart0 receives a small burst of data (~50 bytes) each second.

Any idea what can cause the MSBsrv task to trigger the watchdog?
Could it be the >100ms read on Uart2?

**Comment by Miq1** [comment]
Difficult to tell what does hold up the server task without more details - can you run the ESP32 exception decoder on the abort data to find the spot where the server is at the time the WDT strikes?
The UARTs normally should keep the WDT happy, so it must be something else.

**Comment by bertmelis** [comment]
Are these other tasks in a separate FreeRTOS task? Do these tasks yield?

**Reply by Miq1** [reply]
The offending active task on Core 0 seems to be the MBsrv01RTU task the lib has started. My personal favourite would be the receive function from UART.

**Reply by bertmelis** [reply]
Or sending. There have been some changes to the `flush` function in the Arduino core. (Not so for Platformio but it changed a lot for "regular" Arduino.

**Reply by mstegen** [reply]
Thanks for your replies. There are two tasks, both yield with `vTaskDelay(100 / portTICK_PERIOD_MS);`
So should be enough time to keep the scheduler happy.
I'll do some more tests, see if i can force it to trigger a wdt timeout.

**Reply by mstegen** [reply]
After running fine for a long time (almost 10 days) without any WDT timeouts, it finally triggered the same error again.
This time i had the exception decoder enabled.
Here is the output:


E (839288039) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (839288039) task_wdt:  - IDLE0 (CPU 0)
E (839288039) task_wdt: Tasks currently running:
E (839288039) task_wdt: CPU 0: MBsrv01RTU
E (839288039) task_wdt: CPU 1: loopTask
E (839288039) task_wdt: Aborting.
abort() was called at PC 0x4014ab3c on core 0

ELF file SHA256: 0000000000000000

Backtrace: 0x400887d4:0x3ffbf940 0x40088a51:0x3ffbf960 0x4014ab3c:0x3ffbf980 0x40087045:0x3ffbf9a0 0x400842d5:0x3ffb3930 0x40084fc3:0x3ffb3950 0x40081207:0x3ffb3970 0x4008123c:0x3ffb3990 0x400de2f3:0x3ffb39b0 0x400de38d:0x3ffb39e0 0x400dddd1:0x3ffb3a20 0x40089a62:0x3ffb3ad0
  #0  0x400887d4:0x3ffbf940 in invoke_abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/panic.c:715
  #1  0x40088a51:0x3ffbf960 in abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/panic.c:715
  #2  0x4014ab3c:0x3ffbf980 in task_wdt_isr at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/task_wdt.c:252
  #3  0x40087045:0x3ffbf9a0 in _xt_lowint1 at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/freertos/xtensa_vectors.S:1154
  #4  0x400842d5:0x3ffb3930 in timer_overflow_happened at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/esp_timer_esp32.c:351
      (inlined by) esp_timer_impl_get_time at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/esp_timer_esp32.c:190
  #5  0x40084fc3:0x3ffb3950 in esp_timer_get_time at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/esp_timer.c:435
  #6  0x40081207:0x3ffb3970 in micros at /home/mike/.platformio/packages/framework-arduinoespressif32/cores/esp32/esp32-hal-misc.c:172
  #7  0x4008123c:0x3ffb3990 in delayMicroseconds at /home/mike/.platformio/packages/framework-arduinoespressif32/cores/esp32/esp32-hal-misc.c:172
  #8  0x400de2f3:0x3ffb39b0 in RTUutils::send(HardwareSerial&, unsigned int&, unsigned int, std::function<void (bool)>, unsigned char const*, unsigned short) at .pio/libdeps/release/eModbus/src/RTUutils.cpp:167 (discriminator 1)
  #9  0x400de38d:0x3ffb39e0 in RTUutils::send(HardwareSerial&, unsigned int&, unsigned int, std::function<void (bool)>, ModbusMessage) at .pio/libdeps/release/eModbus/src/RTUutils.cpp:188 (discriminator 2)
  #10 0x400dddd1:0x3ffb3a20 in ModbusServerRTU::serve(ModbusServerRTU*) at .pio/libdeps/release/eModbus/src/RTUutils.h:53
  #11 0x40089a62:0x3ffb3ad0 in vPortTaskWrapper at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/freertos/port.c:355 (discriminator 1)

Rebooting...

I'm not a expert in decoding the exception decoder output, but it looks it's indeed related to the `RTUutils::send` function

**Reply by Miq1** [reply]
Thanks for the exception dump.

The spot it points to is a bit strange for being a reason for a WDT timeout, though:
![grafik](https://user-images.githubusercontent.com/25884124/149180438-6142161e-4444-42db-916d-3c1228ec03d5.png)
The exception would have happened inside ``delayMicroseconds()``, that belongs to the core functions and is supposed to yield to the WDT internally.  The dump then directly is pointing into the Espressif IDF code at ``esp_timer.c``.

So I am afraid I will not be able to do anything about this... 😢 The only thing I can imagine could have happened ist that the calculated time to be spent by ``delayMicroseconds()`` was larger than can fit. If ``last_micros`` was really huge, we may have an overflow at the subtraction from micros(). but we are using ``uint32_t`` here throughout, and any value will suit the ``delayMicroseconds()`` call.

You might have a chance to get a solution if you will post this at the arduino-esp32 project on Github. There are the experts able to look into the IDF code to see what really is happening there.

**Reply by mstegen** [reply]
Thanks Miq1 for your quick reply.

Looking into the `micros()` function, i see that it returns a `uint64_t`.
at line 183 `lastMicros` is set to the output of `micros()`
but `lastMicros` is only 32 bit, and so after 4294 seconds it will only capture the 32 LSB's
Shouldn't `lastMicros` be an `uint64_t` as well?

**Reply by Miq1** [reply]
If you will refer to [this issue in the arduino-esp32 core project](https://github.com/espressif/arduino-esp32/issues/384), you will see that for compatibility reasons the size of millis() and micros() values has been fixed to uint32, despite the fact the ESPs could support 64 bit unsigned longs.
Where did you find the micros() definition with uint64_t?

Edit: wait. I now found where me_no_dev in fact changed the return types to unsigned long, while keeping the parameter values for delay and delayMicroseconds to uint32_t. Slightly inconsistent, as far as I am concerned.

But even given that - what would happen? The if clause either will be true or false, for whatever reason. If it evaluates to true, delayMicroseconds is called with whatever value again, all being valid, even if not sensible. No reason at all to stall the task.

**Reply by mstegen** [reply]
in esp32-hal-misc.c at line 134 they define micros()
they call `esp_timer_get_time()` which returns a `uint64_t`

Yes it would default to true, so it would not cause this issue.. however your code at line 167 would not respect the interval correctly i think?

If you take a look at the delayMicroseconds() function, they also use `uint32_t` to store the value of `micros()` 
This can cause issues..

they seem to have changed it in the 'master' branch now:
https://github.com/espressif/arduino-esp32/blob/master/cores/esp32/esp32-hal-misc.c#L178

here they use uint64_t ...
found the commit...
https://github.com/espressif/arduino-esp32/commit/864597198196c8dd2fed999daf02e4f889238cbc

**Reply by Miq1** [reply]
I eventually will change that in the code. I hope I will not trigger an issue elsewhere. From the Arduino world I knew by heart millis() is uint32_t, so I never considered it could be different in the ESP Arduino core...

But still your issue must have been caused by something different.

**Reply by mstegen** [reply]
As gin66 pointed out:

> Assume m=micros() = 0xffff0000 and us = 0xffff.
> So the calculated value will be e=0xffffffff and no overflow has occurred.
> Consequently the second while loop will be executed.
> If now an interrupt or a high priority task squeezes inbetween and returns after a couple of us, then the micros() may have wrapped around. So it will need wait 4 billion us = 4000s, which could trigger a watchdog.

I just changed the delaymicroseconds() code to use 64 bits. I'll give it another try. Thanks for your help.

**Reply by Miq1** [reply]
De nada! 😁 
Thank you for hinting me to it.

**Reply by Miq1** [reply]
@bertmelis please take note - this may have consequences for the ESP8266 as well.

**Reply by bertmelis** [reply]
I don't think there is any delay involved in async because that would cause watchdog errors. But I'll take a look.

**Reply by Miq1** [reply]
I just found some variables that are assigned a millis() value - just in case...

**Comment by Miq1** [comment]
I made the relevant changes in the non-AsyncTCP code to the ``unsigned long`` data type in the ``master`` branch. 

I have no working test environment at the moment, so I can only state it is compiling without errors. 

If you like, you could give it a try... 😜

**Reply by mstegen** [reply]
Thanks, I'll give it a try.
